### DISCUSSION TITLE
🐛 delta breaks git log

### DISCUSSION THREAD

**Discussion by neo-apz** [dandavison/delta#1387] [General]
I'm experiencing an issue when running `git log` that causes the command to keep producing output and because there are many commits on my branch, the command just keeps printing. It seems no pagination is happening. Everything else works great and only this is causing an issue.

**Comment by dandavison** [comment]
Hi @neo-apz, delta tries to invoke a pager automatically. By default it tries to use `less`, but if you have an environment variable set such as `PAGER=` or `PAGER=cat` then you'd see the behaviour you describe. Can you post

- your OS
- `which less`
- `env | grep -i pager`

I'm converting this to a Discussion because it's not a bug or feature request.

**Reply by neo-apz** [reply]
Hi @dandavison, 

Thanks for the prompt response.

Here are the things you asked for:

- My OS is Oracle Linux 7.
- Initially I was using the default `less` which was at `/usr/bin/less` and it was version `458` but when I saw this issue, I tried manually building the latest less from source (`608`) and  now I'm using that which is at `/usr/local/bin/less` but even with that I have the same problem.
- This is empty (is it normal?)

**Reply by dandavison** [reply]
Hm. A few things you could try to shed light:

Do these page as expected?

```
git log | delta
git log | delta --paging=always
git log | delta --pager=$(which less)
```

These are slightly random tricks to check which less is being invoked by the git process:
```
GIT_PAGER='which less' git log
GIT_PAGER='less --version' git log
```

**Reply by neo-apz** [reply]
I tried all three `git log` commands and they all worked fine (I had disabled `delta` in my `.gitconfig` file before running them) and the two follow-up commands pointed to the latest `less` version that I installed (so `/usr/local/bin/less` and `608`) was shown.

Then I re-enabled `delta` in my config file and reran things and now suddenly everything works fine and I have no idea how/why it's working fine now. 😄 🤷‍♂️

**Reply by dandavison** [reply]
OK cool! All's well that ends well.

**Comment by patrickhousley** [comment]
In case you are coming here because you are on MacOS and when you run `git log`, the log just disappears, use brew to update `less`.

OS: macOS 14.4.1 (23E224)

Before:

```
> less --version
less 581.2 (POSIX regular expressions)
Copyright (C) 1984-2021  Mark Nudelman
```

After:

```
> less --version
less 643 (PCRE2 regular expressions)
Copyright (C) 1984-2023  Mark Nudelman
```

**Comment by scarf005** [comment]
hi @neo-apz, sorry for bothering you. I'm getting the same issue from `delta` installed from [distrobox](https://distrobox.it). the pagination feature only works when i'm inside distrobox, and it doesn't work outside distrobox.

## Reproduction

```sh
$ distrobox create --name fedora --image fedora
$ distrobox enter fedora
[distrobox] sudo dnf install git-delta
[distrobox] distrobox-export --bin $(which delta)
[distrobox] exit
$ which delta
/home/scarf/.local/bin/delta
```

## Situation

```
GIT_PAGER=less git log
```
pagination works normally using `less`;

> Do these page as expected?
> 
> ```
> git log | delta
> git log | delta --paging=always
> git log | delta --pager=$(which less)
> ```

all the three above only worked inside distrobox. it did not work outside distrobox.

> These are slightly random tricks to check which less is being invoked by the git process:
> 
> ```
> GIT_PAGER='which less' git log
> GIT_PAGER='less --version' git log
> ```

the output was same for both in and outside distrobox.

```
$ GIT_PAGER='which less' git log
/usr/bin/less
$ GIT_PAGER='less --version' git log
less 643 (POSIX regular expressions)
Copyright (C) 1984-2023  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: https://greenwoodsoftware.com/less
```

## System

- inside distrobox: fedora 40 (container image)
- outside distrobox: kinoite 40

**Reply by th1000s** [reply]
But plain `git log` and `git log | less` paginate correctly I assume? What less-specific env vars are set, can you check `env | grep -i less`? It might also be an issue with the terminal emulator, but usually `less` warns in this case, e.g. `TERM=dumb git log`

**Reply by scarf005** [reply]
> But plain `git log` and `git log | less` paginate correctly I assume? 

- `git log` without `core.pager` set works.
- `git log` with `core.pager`set to `delta` no longer paginates.
- `git log | less` works.

>  What less-specific env vars are set, can you check env | grep -i less?

```
env | grep -i less
LESSOPEN=||/usr/bin/lesspipe.sh %s
```

>  It might also be an issue with the terminal emulator, but usually less warns in this case, e.g. TERM=dumb git log

- currently using [konsole](https://konsole.kde.org).
- `TERM=dumb git log` gives:

```
scarf@fedora /r/m/h/s/r/aoc-scala (main)> TERM=dumb git log
WARNING: terminal is not fully functional
Press RETURN to continue 
```
![image](https://github.com/user-attachments/assets/70c7d0aa-bea6-4261-9968-1279ce785ea3)

then regular less page with pagination.

**Comment by th1000s** [comment]
Can you `strace` the git and delta calls to `less`, writing it to a log file `-o`, including full syscall args and strings up to a length of 99 `-v -s 99`, following (process) forks `-f`, only looking at successful syscalls for process management `-z -e process`:

`strace -o git.log -v -s 99 -f -z -e process git log -10 --patch`  # without the pager set

vs.

`git log -10 --patch | strace -o delta.log -v -s 99 -f -z -e process delta`

There should be an `execve(..)` call to `less` in both logs, but ignore the one with `--version` in delta. It has the arguments (those are interesting) and the env vars (do *not* post these, they might contain sensitive information), for example:

`execve("/usr/bin/less", ["less", "--RAW-CONTROL-CHARS", "--quit-if-one-screen"], ["COLUMNS=165", "SECRET=01234", /* more env vars */ ]`

`delta` will add `LESSUTFCHARDEF`, `LESSCHARSET` and `LESSANSIENDCHARS` env vars. Do you notice any difference in arguments or other env vars? Does setting the delta env vars in a shell before calling plain `git log` change the behavior?

**Reply by scarf005** [reply]
sorry, had little time, could only post logs.

```sh
scarf@fedora /r/m/h/s/r/aoc-scala (main)> PAGER='' strace -o git.log -v -s 99 -f -z -e process git log -10 --patch 2&> no-pager.log^C
scarf@fedora /r/m/h/s/r/aoc-scala (main)> git log -10 --patch | strace -o delta.log -v -s 99 -f -z -e process delta 2&> delta.log

```

[no-pager.log](https://github.com/user-attachments/files/17105589/no-pager.log)
[delta.log](https://github.com/user-attachments/files/17105594/delta.log)

p.s should this be converted into an issue as it's getting longer?

**Reply by scarf005** [reply]
related: https://github.com/89luca89/distrobox/issues/1572

**Reply by th1000s** [reply]
Well, the log is written to twice, only the strace output is interesting `-o delta.log .. 2&> delta.log` but that was overwritten (Luckily! There might be sensitive data in the env vars!).
